src/compas/datastructures/tree/tree.py

```diff
@@ -0,0 +1,373 @@
+        return self._parent
+
+    @property
+    def children(self):
+        return self._children
+
+    @property
+    def tree(self):
+        if self.is_root:
+            return self._tree
+        else:
+            return self.parent.tree
+
+    @property
+    def data(self):
+        return {
+            "name": self.name,
+            "attributes": self.attributes,
+            "children": [child.data for child in self.children],
+        }
+
+    @classmethod
+    def from_data(cls, data):
+        node = cls(data["name"], data["attributes"])
+        for child in data["children"]:
+            node.add(cls.from_data(child))
+        return node
+
+    def add(self, node):
+        """
+        Add a child node to this node.
+
+        Parameters
+        ----------
+        node : :class:`~compas.datastructures.TreeNode`
+            The node to add.
+
+        Returns
+        -------
+        None
+
+        Raises
+        ------
+        TypeError
+            If the node is not a :class:`~compas.datastructures.TreeNode` object.
+
+        """
+        if not isinstance(node, TreeNode):
+            raise TypeError("The node is not a TreeNode object.")
+        self._children.add(node)
+        node._parent = self
+
+    def remove(self, node):
+        """
+        Remove a child node from this node.
+
+        Parameters
+        ----------
+        node : :class:`~compas.datastructures.TreeNode`
+            The node to remove.
+
+        Returns
+        -------
+        None
+
+        """
+        self._children.remove(node)
+        node._parent = None
+
+    @property
+    def ancestors(self):
+        this = self
+        while this:
+            yield this
+            this = this.parent
+
+    @property
+    def descendants(self):
+        for child in self.children:
+            yield child
+            for descendant in child.descendants:
+                yield descendant
+
+    def traverse(self, strategy="preorder"):
+        """
+        Traverse the tree from this node.
+
+        Parameters
+        ----------
+        strategy : str, optional
+            The traversal strategy. Options are ``"preorder"`` and ``"postorder"``.
+            Default is ``"preorder"``.
+
+        Yields
+        ------
+        :class:`~compas.datastructures.TreeNode`
+            The next node in the traversal.
+
+        Raises
+        ------
+        ValueError
+            If the strategy is not ``"preorder"`` or ``"postorder"``.
+
+        """
+        if strategy == "preorder":
+            yield self
+            for child in self.children:
+                for node in child.traverse(strategy):
+                    yield node
+        elif strategy == "postorder":
+            for child in self.children:
+                for node in child.traverse(strategy):
+                    yield node
+            yield self
+        else:
+            raise ValueError("Unknown traversal strategy: {}".format(strategy))
```

Comment:
preorder and postorder is the same as depth-first, breadth-first?

Reply:
I thought they are, but just realized I got it confused too... Turned out pre and post order are both subcategory of depth-first, the difference is whether you visit current node first or not. None of them are breath-first. Will make some adjustments here

Reply:
@tomvanmele ok now there are two arguments `strategy` and `order`, strategy refers to depth first of breath first, and order is pre or post order (only for depth first). This should better aligns the terminologies.